moto/cloudformation/parsing.py

```diff
@@ -676,10 +668,22 @@ def update(self, template, parameters=None):
             )
             self._parsed_resources.pop(logical_name)
 
+        self._template = template
+        if parameters:
+            self.input_parameters = parameters
+        self.load_mapping()
+        self.load_parameters()
+        self.load_conditions()
+
+        self._resource_json_map = template["Resources"]
+
+        for logical_name in resource_names_by_action["Add"]:
```

Comment:
What does this for-loop do?

Reply:
The `__getitem__` method is lazy and creates resources if they do not yet exist. It has the same initialization steps that were in the original loop (original lines 657-662). This was an attempt to DRY that out and use only one set of initialization steps.

It looks pretty silly though.

I could refactor the initialization out of `__getitem__` and into its own method, then call that method here, if that makes more sense.

Reply:
Ah, right, I see. It does look pretty silly, but I guess that's just because it is a non-intuitive way of resource creation in the first place.
I'm OK with keeping it like this - just put a comment there to explain it, please, to make it easier for whoever comes across this code next.

Reply:
Doneski. https://github.com/spulec/moto/pull/4200/commits/0f7aed7ecf8b124d3274e329a3826da0bf806a5a

I am still trying to understand it, but I am starting to think that `update` should not be trying to initialize added resources at all, given the lazy nature of resources in ResourceMap. Perhaps a prior author of `update` did not want it to be lazy or did not code it correctly to be lazy? Not sure. I will look to follow up with another PR to do something about it.

Reply:
As far as I know, it's pretty close on par with how AWS behaves. And given the complexity of it all, I'm just happy that it works 🙂